torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -1546,6 +1608,28 @@ def _prep_grads_for_backward(self) -> None:
                 p.grad.size() != p._orig_size  # type: ignore[attr-defined]
                 or p.grad.device != p.device
             ):
+                offloaded: bool = p.grad.device != p.device
+                if offloaded:
+                    assert self.cpu_offload.offload_params, \
+                        "`p.grad.device` and `p.device` should be the same " \
+                        "if not offloading parameters to CPU"
+                prev_iter_outside_no_sync: bool = \
+                    p.grad.size() == p._local_shard.shape  # type: ignore[attr-defined]
+                # As long as the previous iteration was outside `no_sync()`,
+                # then we must save the gradient in `_saved_grad_shard`, even
+                # if the current iteration is inside `no_sync()`. This is to
+                # prepare for the next iteration outside `no_sync()`, which may
+                # try to accumulate gradients. FSDP accumulates gradients in
+                # the separate variable `p._saved_grad_shard` to leave `p.grad`
+                # for the per-iteration gradient.
+                if prev_iter_outside_no_sync:
```

Comment:
Since this PR is getting a bit cluttered, I wanted to specifically point this part out. Previously, I presented the logic here incorrectly, but hopefully this should make sense now.

I think the precise condition of when to use `p._saved_grad_shard` is if the previous iteration was outside `no_sync()`.

Suppose we have
(1) some iterations outside `no_sync()` ->
(2) some iterations inside `no_sync()` ->
(3) one iteration outside `no_sync()`. 

- In the pre-backward hook of (3), the FSDP instance holds an unsharded gradient in `p.grad`, which is the result of accumulating gradients from (2).
- It computes that iteration's gradient, which is accumulated with the existing `p.grad` from (2) via the autograd engine and still stored in `p.grad`.
- In the post-backward hook, it reduce-scatters that accumulated gradient stored in `p.grad`.
- After the reduce-scatter, it accumulates the accumulated gradient from (2) and (3) with the accumulated gradient from (1) saved in `_saved_grad_shard`. This "super-accumulated" gradient is stored in `_saved_grad_shard`.
    - This step shows why as long as the previous iteration was outside `no_sync()`, there may be a gradient to accumulate on the first future iteration also outside `no_sync()`.

I had misunderstood the comments from Fairscale (see [here](https://github.com/facebookresearch/fairscale/blob/40e7450f6fd1ba27a71c58da2b49cd19a46b7678/fairscale/nn/data_parallel/fully_sharded_data_parallel.py#L2000-L2003)). I did not realize that the conditioning on inside/outside `no_sync()` referred to the **previous** iteration.